File: backend/src/web/WebRequests.cpp
#include "WebRequests.hpp"
#include <QTimer>
#include <QFile>
#include <QDir>
#include <QDateTime>
#include <QDebug>
#include <QUrlQuery>

WebRequestsLogic::WebRequestsLogic(const QJsonObject &config, QObject *parent)
    : QObject(parent), m_config(config) {
    
    m_netManager = new QNetworkAccessManager(this);
    
    // Parse configuration
    m_baseUrl = config.value("base_url").toString();
    m_requestList = config.value("requests").toArray();
    m_actionList = config.value("actions").toArray();
}

Q_SLOT void WebRequestsLogic::stop() {
    m_isRunning = false;
    emit onStatus("Cancellation pending...");
}

void WebRequestsLogic::waitSeconds(double seconds) {
    QEventLoop loop;
    QTimer::singleShot(static_cast<int>(seconds * 1000), &loop, &QEventLoop::quit);
    loop.exec();
}

QUrl WebRequestsLogic::encodeUrl(const QString &path) const {
    // Correctly join path to base URL
    QUrl url = m_baseUrl;
    if (!path.isEmpty()) {
        url = url.resolved(QUrl(path));
    }
    return url;
}

QByteArray WebRequestsLogic::createPostData(const QString &paramStr) const {
    // Converts 'key:val, k2:v2' into QUrlQuery for POST data (application/x-www-form-urlencoded)
    if (paramStr.isEmpty()) {
        return QByteArray();
    }
    
    QUrlQuery postData;
    QStringList pairs = paramStr.split(',');
    
    for (const QString &pair : pairs) {
        if (pair.contains(':')) {
            QStringList parts = pair.split(':', Qt::SkipEmptyParts);
            if (parts.size() == 2) {
                postData.addQueryItem(parts[0].trimmed(), parts[1].trimmed());
            }
        } else {
            // Emit error if format is invalid
            emit onError(QString("Invalid POST data format: '%1'. Skipping.").arg(pair));
        }
    }
    return postData.toString(QUrl::FullyEncoded).toUtf8();
}

void WebRequestsLogic::runActions(QNetworkReply *reply) {
    // Check for HTTP errors before running actions, mirroring Python's raise_for_status()
    if (reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt() >= 400) {
        return;
    }
    
    for (const QJsonValue &actionValue : m_actionList) {
        if (!m_isRunning) return;

        QJsonObject action = actionValue.toObject();
        QString type = action.value("type").toString();
        QString param = action.value("param").toString();

        try {
            if (type == "Print Response URL") {
                emit onStatus(QString("  > Action: Response URL: %1").arg(reply->url().toString()));
            }
            else if (type == "Print Response Status Code") {
                int statusCode = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
                emit onStatus(QString("  > Action: Status Code: %1").arg(statusCode));
            }
            else if (type == "Print Response Headers") {
                QString headersStr;
                for (const auto &pair : reply->rawHeaderPairs()) {
                    headersStr += QString("    %1: %2\n").arg(QString(pair.first), QString(pair.second));
                }
                emit onStatus("  > Action: Response Headers:\n" + headersStr.trimmed());
            }
            else if (type == "Print Response Content (Text)") {
                // Truncate content
                QString contentPreview = QString(reply->readAll()).trimmed();
                emit onStatus("  > Action: Response Content:\n" + contentPreview);
            }
            else if (type == "Save Response Content (Binary)") {
                if (param.isEmpty()) {
                    emit onError("  > Action: Save failed. No file path provided in parameter.");
                    continue;
                }
                
                QString filepath = param;
                if (QFileInfo(param).isDir()) {
                    // Get filename from URL
                    QString filename = reply->url().fileName().split("?").first();
                    if (filename.isEmpty()) {
                        filename = QString("response_%1.dat").arg(QDateTime::currentMSecsSinceEpoch());
                    }
                    filepath = QDir(param).filePath(filename);
                }
                
                QDir().mkpath(QFileInfo(filepath).absolutePath());
                QFile file(filepath);
                if (file.open(QIODevice::WriteOnly)) {
                    file.write(reply->readAll());
                    file.close();
                    emit onStatus(QString("  > Action: Response content saved to %1").arg(filepath));
                } else {
                    emit onError(QString("  > Action: Failed to save file to %1: %2").arg(filepath, file.errorString()));
                }
            }
        } catch (const std::exception &e) {
            emit onError(QString("  > Action: Failed to execute '%1': %2").arg(type, e.what()));
        }
    }
}

bool WebRequestsLogic::executeRequest(const QJsonObject &req) {
    if (!m_isRunning) return false;

    QString reqType = req.value("type").toString("GET");
    QString param = req.value("param").toString();
    
    QUrl urlToRequest = encodeUrl(reqType == "GET" ? param : QString());
    
    QNetworkRequest request(urlToRequest);
    request.setRawHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.75 Safari/537.36");

    QNetworkReply *reply = nullptr;
    QEventLoop loop;
    
    if (reqType == "GET") {
        emit onStatus(QString("Executing GET: %1").arg(urlToRequest.toString()));
        reply = m_netManager->get(request);
    } 
    else if (reqType == "POST") {
        QByteArray postData = createPostData(param);
        request.setHeader(QNetworkRequest::ContentTypeHeader, "application/x-www-form-urlencoded");
        emit onStatus(QString("Executing POST: %1 with data: %2").arg(urlToRequest.toString(), QString(postData)));
        reply = m_netManager->post(request, postData);
    } else {
        emit onError(QString("Unsupported request type: %1").arg(reqType));
        return false;
    }

    // Connect signals for sequential execution (mimics blocking nature)
    QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
    
    // Start waiting for the reply to finish
    QTimer::singleShot(10000, &loop, &QEventLoop::quit); // Timeout (10 seconds)
    loop.exec();
    
    // Check if the reply finished normally or timed out/was cancelled
    if (!reply->isFinished()) {
        reply->abort();
        emit onError("Request failed: Timeout.");
        reply->deleteLater();
        return false;
    }

    emit onStatus(QString("Request complete. Status: %1").arg(reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt()));

    // Check for errors (Network/HTTP)
    if (reply->error() != QNetworkReply::NoError) {
        if (reply->error() == QNetworkReply::TimeoutError) {
            emit onError("Request failed: Timeout.");
        } else {
            // Check for explicit HTTP errors (4xx/5xx)
            int statusCode = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
            if (statusCode >= 400) {
                 emit onError(QString("Request failed: HTTP %1 %2").arg(statusCode).arg(reply->errorString()));
            } else {
                 emit onError(QString("Request failed: Connection Error: %1").arg(reply->errorString()));
            }
        }
        reply->deleteLater();
        return false;
    }
    
    // Run actions on success
    runActions(reply);
    
    reply->deleteLater();
    return true;
}

void WebRequestsLogic::run() {
    emit onStatus(QString("Starting request sequence for %1").arg(m_baseUrl.toString()));

    for (int i = 0; i < m_requestList.size(); ++i) {
        if (!m_isRunning) {
            emit onStatus("Request sequence cancelled.");
            emit onFinished("Cancelled.");
            return;
        }

        QJsonObject req = m_requestList[i].toObject();
        emit onStatus(QString("--- Request %1/%2: [%3] ---").arg(i + 1).arg(m_requestList.size()).arg(req.value("type").toString("GET")));
        
        executeRequest(req);
        
        waitSeconds(0.5); // Small delay between requests
    }
    
    if (m_isRunning) {
        emit onStatus("--- All requests finished. ---");
        emit onFinished("All requests finished.");
    }
}

File: backend/src/web/web_requests.py
import os
import time
import requests

from PySide6.QtCore import QObject, Signal


class WebRequestsLogic(QObject):
    """
    Performs a sequence of web requests using the 'requests' library
    and emits Qt signals for status updates.
    """

    # === SIGNALS ===
    on_status = Signal(str)
    on_error = Signal(str)
    on_finished = Signal(str)

    def __init__(self, config: dict):
        super().__init__()
        self.config = config
        self.base_url = config.get("base_url")
        self.request_list = config.get("requests", [])
        self.action_list = config.get("actions", [])
        self._is_running = True  # Flag to control cancellation

    def stop(self):
        """Sets the flag to stop the execution loop."""
        self._is_running = False
        self.on_status.emit("Cancellation pending...")

    def _parse_post_data(self, param_str: str) -> dict:
        """Converts 'key:val, k2:v2' into a dictionary."""
        if not param_str:
            return {}

        data_dict = {}
        try:
            pairs = param_str.split(",")
            for pair in pairs:
                if ":" in pair:
                    key, value = pair.split(":", 1)
                    data_dict[key.strip()] = value.strip()
                else:
                    self.on_error.emit(f"Invalid POST data format: '{pair}'. Skipping.")
            return data_dict
        except Exception as e:
            self.on_error.emit(f"Error parsing POST data: {e}")
            return {}

    def _run_actions(self, response: requests.Response):
        """Runs the defined action list on the given response."""

        for action in self.action_list:
            if not self._is_running:
                return  # Stop processing actions if cancelled

            action_type = action.get("type")
            param = action.get("param")

            try:
                if action_type == "Print Response URL":
                    self.on_status.emit(f"  > Action: Response URL: {response.url}")

                elif action_type == "Print Response Status Code":
                    self.on_status.emit(
                        f"  > Action: Status Code: {response.status_code}"
                    )

                elif action_type == "Print Response Headers":
                    headers_str = "\n".join(
                        f"    {k}: {v}" for k, v in response.headers.items()
                    )
                    self.on_status.emit(
                        f"  > Action: Response Headers:\n {headers_str}"
                    )

                elif action_type == "Print Response Content (Text)":
                    # Truncate to avoid flooding the log
                    content_preview = response.text.strip()
                    self.on_status.emit(
                        f"  > Action: Response Content:\n {content_preview}"
                    )

                elif action_type == "Save Response Content (Binary)":
                    if not param:
                        self.on_error.emit(
                            "  > Action: Save failed. No file path provided in parameter."
                        )
                        continue

                    # Check if param is a directory
                    if os.path.isdir(param):
                        # Try to get a filename from the URL
                        try:
                            filename = os.path.basename(response.url.split("?")[0])
                            if not filename:
                                filename = f"response_{int(time.time())}.dat"
                        except Exception:
                            filename = f"response_{int(time.time())}.dat"
                        filepath = os.path.join(param, filename)
                    else:
                        filepath = param

                    try:
                        os.makedirs(os.path.dirname(filepath), exist_ok=True)
                        with open(filepath, "wb") as f:
                            f.write(response.content)
                        self.on_status.emit(
                            f"  > Action: Response content saved to {filepath}"
                        )
                    except Exception as e:
                        self.on_error.emit(
                            f"  > Action: Failed to save file to {filepath}: {e}"
                        )

            except Exception as e:
                self.on_error.emit(
                    f"  > Action: Failed to execute '{action_type}': {e}"
                )

    def run(self):
        """
        Main execution loop.
        Iterates through the request list and performs actions.
        """
        self.on_status.emit(f"Starting request sequence for {self.base_url}")

        for i, req in enumerate(self.request_list):
            if not self._is_running:
                self.on_status.emit("Request sequence cancelled.")
                self.on_finished.emit("Cancelled.")
                return

            req_type = req.get("type", "GET")
            param = req.get("param")

            url_to_request = self.base_url

            self.on_status.emit(
                f"--- Request {i+1}/{len(self.request_list)}: [{req_type}] ---"
            )

            try:
                response = None
                if req_type == "GET":
                    if param:  # Use param as URL suffix
                        url_to_request = os.path.join(self.base_url, param).replace(
                            "\\", "/"
                        )
                    self.on_status.emit(f"Executing GET: {url_to_request}")
                    response = requests.get(
                        url_to_request, timeout=10, allow_redirects=True
                    )

                elif req_type == "POST":
                    post_data = self._parse_post_data(param)
                    self.on_status.emit(
                        f"Executing POST: {url_to_request} with data: {post_data}"
                    )
                    response = requests.post(
                        url_to_request, data=post_data, timeout=10, allow_redirects=True
                    )

                if response is not None:
                    self.on_status.emit(
                        f"Request complete. Status: {response.status_code}"
                    )
                    response.raise_for_status()  # Raise HTTPError for bad responses (4xx or 5xx)

                    # Run all defined actions on this response
                    self._run_actions(response)

            except requests.exceptions.HTTPError as he:
                self.on_error.emit(
                    f"Request failed: HTTP {he.response.status_code} {he.response.reason}"
                )
            except requests.exceptions.ConnectionError as ce:
                self.on_error.emit(f"Request failed: Connection Error: {ce}")
            except requests.exceptions.Timeout as te:
                self.on_error.emit(f"Request failed: Timeout: {te}")
            except Exception as e:
                self.on_error.emit(f"Request failed: An unexpected error occurred: {e}")

            time.sleep(0.5)  # Small delay between requests

        if self._is_running:
            self.on_status.emit("--- All requests finished. ---")
            self.on_finished.emit("All requests finished.")
